pkg/tsdb/cloudwatch/cloudwatch.go

```diff
@@ -71,7 +76,7 @@ type SessionCache interface {
 	GetSession(c awsds.SessionConfig) (*session.Session, error)
 }
 
-func newExecutor(im instancemgmt.InstanceManager, cfg *setting.Cfg, sessions SessionCache, features featuremgmt.FeatureToggles) *cloudWatchExecutor {
+func newExecutor(im instancemgmt.InstanceManager, cfg *setting.Cfg, sessions SessionCache, features featuremgmt.FeatureToggles, logsTimeout ...time.Duration) *cloudWatchExecutor {
```

Comment:
After looking over the whole work, I think you don't need to add the logsTimeout to the executor. Instead you can inject it in your tests through the instance settings.

What you can do is (see my other comment) validate the time duration string and initialize it in `LoadCloudWatchSettings`. (I would add a test for that in settings_test.go)

Reply:
I think you have a great solution!

I originally thought of using the "DTO" solution which is to have one struct type to deserialize the JSON and another struct type to actually use everywhere else in the code, but I like the custom Unmarshaller you have created more.


I have a couple of thoughts:
1. After reading [this stackoverflow](https://stackoverflow.com/questions/48050945/how-to-unmarshal-json-into-durations), I think I slightly prefer using a struct containing time.Duration because of the comment "One thing I noticed when implementing this solution but with a type alias was that you lose some of the access to existing methods on the time.Duration type. So one example was that when implementing MarshalJSON the String method no longer existed." What do you think of that?

2. I am a little undecided if we should now set the defaults in the Unmarshal method on Duration and in the initialization in the CloudWatchSettings struct. Let me know what you think. Something like:

```golang
// logs timeout default is 30 minutes, the same as timeout in frontend logs query
// note: for alerting queries, the context will be cancelled before that unless evaluation_timeout_seconds in defaults.ini is increased (default: 30s)
const defaultLogsTimeout = 30 * time.Minute

type Duration struct {
	time.Duration
}

type CloudWatchSettings struct {
	awsds.AWSDatasourceSettings
	Namespace               string   `json:"customMetricsNamespaces"`
	SecureSocksProxyEnabled bool     `json:"enableSecureSocksProxy"` // this can be removed when https://github.com/grafana/grafana/issues/39089 is implemented
	LogsTimeout             Duration `json:"logsTimeout"`
}

func LoadCloudWatchSettings(config backend.DataSourceInstanceSettings) (CloudWatchSettings, error) {
	instance := CloudWatchSettings{LogsTimeout: Duration{defaultLogsTimeout}}
	if config.JSONData != nil && len(config.JSONData) > 1 {
		if err := json.Unmarshal(config.JSONData, &instance); err != nil {
			return CloudWatchSettings{}, fmt.Errorf("could not unmarshal DatasourceSettings json: %w", err)
		}
	}

	if instance.Region == "default" || instance.Region == "" {
		instance.Region = instance.DefaultRegion
	}

	if instance.Profile == "" {
		instance.Profile = config.Database
	}

	instance.AccessKey = config.DecryptedSecureJSONData["accessKey"]
	instance.SecretKey = config.DecryptedSecureJSONData["secretKey"]

	return instance, nil
}

func (d *Duration) UnmarshalJSON(b []byte) error {
	var logsTimeout interface{}
	if err := json.Unmarshal(b, &logsTimeout); err != nil {
		return err
	}

	switch value := logsTimeout.(type) {
	case float64:
		d.Duration = time.Duration(value)
	case string:
		var err error
		d.Duration, err = time.ParseDuration(value)
		if err != nil {
			d.Duration = defaultLogsTimeout
		}
	default:
		d.Duration = defaultLogsTimeout
	}

	return nil
}
```

Reply:
I didn't know about 1! I will change it to struct to keep the methods in case we need them for later :) 

I'm not sure about 2. either. I put it where the other "defaults" for settings are, and maaaaybe I'd rather keep it there cause I feel like it's more visible there🤔

Reply:
Sounds good!